microbench/src/main/java/io/netty/microbench/concurrent/DefaultPromiseBenchmark.java

```diff
@@ -0,0 +1,162 @@
+    private TestGenericFutureListener[] lastListeners;
+    private static AtomicLong doneCount = new AtomicLong();
+    private static CountDownLatch testLatch;
+    private final Runnable toEventLoopRunnable = new Runnable() {
+        @Override
+        public void run() {
+            startOutEventLoop();
+        }
+    };
+
+    @Param({ "100000", "1000000" })
+    public int numPromises;
+
+    @Param({ "1", "4" })
+    public int numListeners;
+
+    @Param({ "0", "4" })
+    public int lateListeners;
+
+    @AuxCounters
+    @State(Scope.Thread)
+    public static class DoneCounter {
+        @Setup(Level.Iteration)
+        public void clean() {
+            doneCount.set(0);
+        }
+
+        public long doneCount() {
+            return doneCount.get();
+        }
+    }
+
+    @Setup(Level.Trial)
+    public void setupTrial() {
+        executorService = Executors.newSingleThreadExecutor();
+        executor = new DefaultEventExecutor(executorService);
+        listeners = new TestGenericFutureListener[numListeners];
+        for (int i = 0; i < listeners.length; ++i) {
+            listeners[i] = new TestGenericFutureListener();
+        }
+        lastListeners = new TestGenericFutureListener[lateListeners];
+        for (int i = 0; i < lastListeners.length; ++i) {
+            lastListeners[i] = new TestGenericFutureListener();
+        }
+    }
+
+    @Setup(Level.Invocation)
+    public void setupInvocation() {
+        testLatch = new CountDownLatch(numPromises * (listeners.length + lastListeners.length));
+    }
+
+    @TearDown(Level.Trial)
+    public void teardownTrial() {
+        final Future<?> f;
+        try {
+            executorService.shutdown();
+        } finally {
+            f = executor.shutdownGracefully(0, 0, TimeUnit.MILLISECONDS);
+        }
+        try {
+            executorService.awaitTermination(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        try {
+            f.get(5, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        } catch (ExecutionException e) {
+            e.printStackTrace();
+        } catch (TimeoutException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Benchmark
+    public void testInEventLoop(DoneCounter doneCounter) throws Exception {
+        startInEventLoop();
+        testLatch.await();
+    }
+
+    @Benchmark
+    public void testOutEventLoop(DoneCounter doneCounter) throws Exception {
+        startOutEventLoop();
+        testLatch.await();
+    }
+
+    private void startInEventLoop() {
+        executor.execute(toEventLoopRunnable);
+    }
+
+    private void startOutEventLoop() {
+        for (int i = 0; i < numPromises; ++i) {
+            Promise<Void> p = new DefaultPromise<Void>(executor);
+            p.addListeners(listeners);
+            p.setSuccess(null);
+            p.addListeners(lastListeners);
```

Comment:
> Do you have a suggestion to improve this? Executing a Runnable on the event loop seems heavy handed for what is being measured here.

Honestly, this is why I avoid microbenchmarks. It is so difficult to produce useful numbers, and even then I typically can't trust them, and even then I don't know if the change has any real-world impact. I typically try to start with a real use-case (as in, code) and produce a small "end to end" semi-realistic benchmark and go from there. If the optimizations are in the noise, then they don't matter (unless the noise is high; if the noise is 5%, then you have to try alternative approaches or optimize the long tail).

Note that still has non-realistic properties for GC behavior. Real GC degrades as the retained size increases, and in a simple benchmark the retained size is typically small or pointer-less nor is run long enough.

I've had good success running small benchmarks and having gains be visible in actual applications, but only when the work load is computation-based and single-threaded. And in Java even that is hard because of the JIT and it getting performance gains due to unrealistic optimizations in practice.

>> close to a zero chance of completing before the running of listeners when outside the event loop.

> Do you mean when inside the event loop?

No, I mean outside the event loop. Inside the event loop the listeners would be guaranteed to have been executed. Outside the event loop, listener execution has added latency before running, such that `addListeners()` is highly likely to be run before the listeners have been notified.

>> such that a long queue formed.

> A new promise is created on each iteration, so the listener queue should not grow larger than listeners + lastListeners for each promise.

I'm speaking more about time. If a queue forms, then there will be limited interaction between the threads. I don't know which work is cheaper: registering the listeners or notifying them. But if notifying them takes longer, then a queue forms and grows as the test progresses. A 50µs queue would mean no contention between the threads. But we don't know if a queue forms, so there may be interaction between the threads or there may not. Or maybe there is oscillation as a queue forms due to latency in waking the event loop (or contention) then the event loop drains the queue faster than the producer. So we don't actually know what interaction is being benchmarked.